adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/CoverageRDD.scala

```diff
@@ -91,8 +144,25 @@ case class CoverageRDD(rdd: RDD[Coverage],
 
   def union(rdds: CoverageRDD*): CoverageRDD = {
     val iterableRdds = rdds.toSeq
-    CoverageRDD(rdd.context.union(rdd, iterableRdds.map(_.rdd): _*),
-      iterableRdds.map(_.sequences).fold(sequences)(_ ++ _))
+
+    val mergedSequences = iterableRdds.map(_.sequences).fold(sequences)(_ ++ _)
+
+    if (iterableRdds.forall(rdd => rdd match {
+      case DatasetBoundCoverageRDD(_, _) => true
+      case _                             => false
+    })) {
+      DatasetBoundCoverageRDD(iterableRdds.map(_.dataset)
+        .fold(dataset)(_.union(_)), mergedSequences)
+    } else {
+      RDDBoundCoverageRDD(rdd.context.union(rdd, iterableRdds.map(_.rdd): _*),
```

Comment:
If you sample from each `rdd` and partition accordingly, you can reduce the skew. No matter what you're going to have to repartition data if you want sorted data after the merge. You may as well rebalance the data at the same time, I suppose. However, now we are talking about merging sorted data, rather than simply a union. I have unintentionally blurred the line between them. We can open an issue if it's something we want to do in the future.